libvmi/os/linux/elfparse.c

```diff
@@ -0,0 +1,422 @@
+        return VMI_SUCCESS;
+
+    uint32_t buckets[table[0]];
+    _ctx.addr += sizeof(bloom);
+    if (VMI_FAILURE == vmi_read(vmi, &_ctx, sizeof(buckets), buckets, NULL))
+        return VMI_FAILURE;
+
+    uint32_t symix = buckets[hash % table[0]];
+    if (symix < table[1])
+        return VMI_FAILURE;
+
+    for (_ctx.addr += sizeof(buckets) + sizeof(uint32_t) * (symix - table[1]);;
+            symix++, _ctx.addr += sizeof(uint32_t)) {
+        uint32_t chain;
+        if (VMI_FAILURE == vmi_read_32(vmi, &_ctx, &chain))
+            break;
+
+        if ((hash & ~0b1) == (chain & ~0b1)) {
+            struct elf64_sym sym;
+            __ctx.addr = sym_tab + sizeof(struct elf64_sym) * symix;
+            if (VMI_FAILURE == vmi_read(vmi, &__ctx, sizeof(struct elf64_sym), &sym, NULL))
+                break;
+
+            /*
+             * technically we can still have hash collisions here,
+             * so we cross-check our result with the string table.
+             */
+            __ctx.addr = str_tab + sym.st_name;
+            char *n = vmi_read_str(vmi, &__ctx);
+            if (!n || strcmp(name, n) != 0) {
+                if (n)
+                    free(n);
+                continue;
+            }
+
+            free(n);
+            *rva = sym.st_value;
+            return VMI_SUCCESS;
+        }
+
+        if (chain & 0b1)
+            break;
+    }
+
+    return VMI_FAILURE;
+}
+
+static uint32_t gnu_hash_max_symbols(
+    vmi_instance_t vmi,
+    const access_context_t *ctx,
+    addr_t hash_tab,
+    uint32_t *max_symbols)
+{
+    access_context_t _ctx = *ctx;
+    uint32_t table[4];
+
+    _ctx.addr = hash_tab;
+    if (VMI_FAILURE == vmi_read(vmi, &_ctx, sizeof(table), table, NULL))
+        return VMI_FAILURE;
+
+    uint32_t buckets[table[0]];
+    _ctx.addr += sizeof(table) + sizeof(uint64_t) * table[2];
+    if (VMI_FAILURE == vmi_read(vmi, &_ctx, sizeof(buckets), buckets, NULL))
+        return VMI_FAILURE;
+
+    uint32_t last = 0;
+    for (uint32_t i = 0; i < table[0]; i++)
+        if (buckets[i] > last)
+            last = buckets[i];
+
+    for (;; last++) {
+        uint32_t chain;
+        _ctx.addr = hash_tab + sizeof(table) + sizeof(uint64_t) * table[2]
+                    + sizeof(buckets) + sizeof(uint32_t) * (last - table[1]);
+        if (VMI_FAILURE == vmi_read_32(vmi, &_ctx, &chain))
+            return VMI_FAILURE;
+
+        if (chain & 0b1)
+            break;
+    }
+
+    *max_symbols = last;
+    return VMI_SUCCESS;
+}
+
+/* returns the rva value for a linux ELF dynamic symbol */
+status_t
+linux_export_to_rva(
+    vmi_instance_t vmi,
+    const access_context_t *ctx,
+    const char *symbol,
+    addr_t *rva)
+{
+    addr_t hash = 0, gnu_hash = 0, str = 0, sym = 0;
+    uint8_t image[1024] = { 0 };
+
+    if (VMI_FAILURE == elfparse_get_image(vmi, ctx, sizeof(image), image))
+        return VMI_FAILURE;
+
+    elfparse_assign_tables(vmi, ctx, image, &hash, &gnu_hash, &str, &sym);
```

Comment:
After a bit of testing it seems the stack protector trips because of the number of arguments being passed to this function. If you can reduce the number of inputs so that the function can take all inputs via registers the warning disappears. So for example instead of passing individual pointers for `&hash, &gnu_hash, &str, &sym`, pass a pointer to a structure with those fields, or simply a pointer to an array.